org.eclipse.xtext.xbase.junit/src/org/eclipse/xtext/xbase/compiler/InMemoryJavaCompiler.java

```diff
@@ -0,0 +1,274 @@
+							@Override
+							protected URLConnection openConnection(URL it) throws IOException {
+								return new URLConnection(it) {
+									@Override
+									public void connect() {
+									}
+
+									@Override
+									public InputStream getInputStream() {
+										return new ByteArrayInputStream(bytes);
+									}
+								};
+							}
+						});
+					}
+				}
+				return null;
+			} catch (IOException e) {
+				throw Exceptions.sneakyThrow(e);
+			}
+		}
+	}
+
+	public static class Result {
+		private final Set<CategorizedProblem> compilationProblems = new LinkedHashSet<>();
+
+		private final HashMap<String, byte[]> classMap = new HashMap<String, byte[]>();
+
+		private final ClassLoader parentClassLoader;
+
+		public ClassLoader getClassLoader() {
+			return new InMemoryJavaCompiler.ByteClassLoader(classMap, parentClassLoader);
+		}
+
+		public Result(ClassLoader parentClassLoader) {
+			this.parentClassLoader = parentClassLoader;
+		}
+
+		public Set<CategorizedProblem> getCompilationProblems() {
+			return compilationProblems;
+		}
+	}
+
+	private final INameEnvironment nameEnv;
+
+	private final ClassLoader parentClassLoader;
+
+	private final CompilerOptions compilerOptions;
+
+	public InMemoryJavaCompiler(ClassLoader parent, JavaVersion javaVersion) {
+		this.nameEnv = new InMemoryJavaCompiler.ClassLoaderBasedNameEnvironment(parent);
+		this.parentClassLoader = parent;
+		this.compilerOptions = new CompilerOptions();
+		this.setSourceLevel(toClassFmt(javaVersion));
+		this.setComplianceLevel(toClassFmt(javaVersion));
+		this.compilerOptions.targetJDK = toClassFmt(javaVersion);
+		this.compilerOptions.inlineJsrBytecode = true;
+		this.compilerOptions.preserveAllLocalVariables = true;
+	}
+
+	public InMemoryJavaCompiler(ClassLoader parent, CompilerOptions compilerOptions) {
+		this.nameEnv = new InMemoryJavaCompiler.ClassLoaderBasedNameEnvironment(parent);
+		this.parentClassLoader = parent;
+		this.compilerOptions = new CompilerOptions(compilerOptions.getMap());
+	}
+
+	private long toClassFmt(JavaVersion version) {
+		return version.toJdtClassFileConstant();
+	}
+
+	/**
+	 * sets the source level (see @link(org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants))
+	 */
+	private void setSourceLevel(long jdkVersion) {
+		try {
+			compilerOptions.sourceLevel = jdkVersion;
+			try {
+				CompilerOptions.class.getField("originalSourceLevel").setLong(compilerOptions, jdkVersion);
+			} catch (NoSuchFieldException e) {
+				// ignore
+			}
+		} catch (IllegalAccessException e) {
+			throw Exceptions.sneakyThrow(e);
+		}
+	}
+
+	/**
+	 * sets the compliance level (see @link(org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants))
+	 */
+	private void setComplianceLevel(long jdkVersion) {
+		try {
+			compilerOptions.complianceLevel = jdkVersion;
+			try {
+				CompilerOptions.class.getField("originalComplianceLevel").setLong(compilerOptions, jdkVersion);
+			} catch (NoSuchFieldException e) {
```

Comment:
// these fields have been introduces in JDT 3.7

Comment is lost and probably relevant.